assignment5/AlphabetDiscovering.java

```diff
@@ -0,0 +1,107 @@
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /*
+     * Class to solve the problem, which does topological sort in existing graph
+     */
+    public static class Ordering {
+        private final Map<Character, Vertex> vertexMap = new HashMap<>();
+
+        /*
+         * Method to start the process of ordering. It scans the given wordlist and creates a graph
+         * --------------------------------------------------------
+         * @param words - given dictionary in unknown language
+         * --------------------------------------------------------
+         * @return initialised Ordering object with a graph
+         */
+        public static Ordering seedWords(List<String> words) {
+            Ordering o = new Ordering();
+            LetterScanner letterScanner = new LetterScanner();
+            letterScanner.scan(words, o);
+            return o;
+        }
+
+        /*
+         * Method to do topological sort of the graph
+         * ------------------------------------------
+         * @return list of the found characters - unknown alphabet
+         */
+        public List<Character> topologicalOrdering() {
+            Set<Vertex> unmarked = new HashSet<>(vertexMap.values());
+            Set<Vertex> visiting = new HashSet<>();
+            Set<Vertex> visited = new HashSet<>();
+
+            List<Character> ordered = new LinkedList<>();
+            while (!unmarked.isEmpty()) {
+                Vertex v = unmarked.iterator().next();
+                visitVertex(v, unmarked, visiting, visited, ordered);
+            }
+
+            Collections.reverse(ordered);
+            return ordered;
+        }
+
+        /*
+         * Method, which does the main part of topological sort.
+         * It visits vertexes of the given graph.
+         * ------------------------------------------------
+         * @param v - vertex to visit
+         * @param unmarked - set of unmarked vertex
+         * @param visiting - set of vising vertex
+         * @param visited - set of visited vertex
+         * @ordered - list of ordered characters
+         */
+        private static void visitVertex(Vertex v, Set<Vertex> unmarked, Set<Vertex> visiting,
+                                  Set<Vertex> visited, List<Character> ordered) {
+            if (visiting.contains(v)) {
+                throw new RuntimeException("Cyclic graph. No ordering possible.");
+            }
+
+            if (unmarked.remove(v)) {
+                visiting.add(v);
+                for (Vertex n : v.getNeighbours()) {
+                    visitVertex(n, unmarked, visiting, visited, ordered);
+                }
+                visiting.remove(v);
+                visited.add(v);
+                ordered.add(v.getValue());
+            }
+        }
+
+        public void onOrderedPair(char u, char v) {

```

Comment:
Does not need to be public - clients that want to discover alphabet do not need to use it, so it's better to hide it. From language perspective it could be private (LetterScanner will have access to it). You could also leave it in package scope (i.e. without public/private modifier).